src/coins.cpp

```diff
@@ -121,24 +121,29 @@ CCoinsModifier CCoinsViewCache::ModifyCoins(const uint256 &txid) {
  * outputs from a transaction.  It assumes that BIP 30 (no duplicate txids)
  * applies and has already been tested for (or the test is not required due to
  * BIP 34, height in coinbase).  If we can assume BIP 30 then we know that any
- * transaction we are adding to the UTXO must not already exist in the utxo
- * unless it is fully spent.  Thus we can check only if it exists at the current
- * level of the cache, in which case it is not safe to mark it FRESH (b/c then
- * spentness might still need to flushed) and if it doesn't exist in the current
- * cache, we know it either doesn't exist or is pruned in parent caches, which
- * is the definition of FRESH.  The exception to this is the two historical
- * violations of BIP 30 in the chain, both of which were coinbases.  We do not
- * mark these fresh so we we can ensure that they will still be properly
- * overwritten when spent.
+ * non-coinbase transaction we are adding to the UTXO must not already exist in
+ * the utxo unless it is fully spent.  Thus we can check only if it exists DIRTY
+ * at the current level of the cache, in which case it is not safe to mark it
+ * FRESH (b/c then its spentness still needs to flushed).  If it's not dirty and
+ * doesn't exist or is pruned in the current cache, we know it either doesn't
+ * exist or is pruned in parent caches, which is the definition of FRESH.  The
+ * exception to this is the two historical violations of BIP 30 in the chain,
+ * both of which were coinbases.  We do not mark these fresh so we we can ensure
+ * that they will still be properly overwritten when spent.
  */
 CCoinsModifier CCoinsViewCache::ModifyNewCoins(const uint256 &txid, bool coinbase) {
     assert(!hasModifier);
     std::pair<CCoinsMap::iterator, bool> ret = cacheCoins.insert(std::make_pair(txid, CCoinsCacheEntry()));
     if (!coinbase) {
-        if (ret.second) {
-            ret.first->second.flags = CCoinsCacheEntry::FRESH;
-        } else {
-            assert(ret.first->second.coins.IsPruned());
+        // New coins must not already exist.
+        assert(ret.first->second.coins.IsPruned());
+
+        if (!(ret.first->second.flags & CCoinsCacheEntry::DIRTY)) {
+            // If the coin is known to be pruned (have no unspent outputs) in
+            // the current view and the cache entry is not dirty, we know the
+            // coin also must be pruned in the parent view as well, so it is safe
+            // to mark this fresh.
+            ret.first->second.flags |= CCoinsCacheEntry::FRESH;
         }
     }
     ret.first->second.coins.Clear();
```

Comment:
Which line?  The `Clear()`?  Isn't that needed for the historical coinbase overwrite cases?